exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/LimitUnionExchangeTransposeRule.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.planner.physical;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.drill.exec.planner.logical.RelOptHelper;
+
+import java.math.BigDecimal;
+
+public class LimitUnionExchangeTransposeRule extends Prule{
+  public static final RelOptRule INSTANCE = new LimitUnionExchangeTransposeRule();
+
+  private LimitUnionExchangeTransposeRule() {
+    super(RelOptHelper.some(LimitPrel.class, RelOptHelper.any(UnionExchangePrel.class)), "LimitUnionExchangeTransposeRule");
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    final LimitPrel limit = (LimitPrel) call.rel(0);
+
+    return !limit.isPushDown();
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    final LimitPrel limit = (LimitPrel) call.rel(0);
+    final UnionExchangePrel unionExchangePrel = (UnionExchangePrel) call.rel(1);
+
+    RelNode child = unionExchangePrel.getInput();
+
+    final int offset = limit.getOffset() != null ? Math.max(0, RexLiteral.intValue(limit.getOffset())) : 0;
+    final int fetch = limit.getFetch() != null?  Math.max(0, RexLiteral.intValue(limit.getFetch())) : 0;
```

Comment:
Thanks, Julian!

Good point about the case of fetch == null.  When fetch == null, the query should return all the remaining rows starting from offset. In such case, seems it does not make sense to push Limit operator, as the underline operator has to go through all the records.

I modified the rule, so that it will not push limit when fetch == null. I also made change to the row count estimation for the case of fetch == null.

For Sean's comment, this patch is for a special case in Drill's physical planing to handle limit on top of UnionExchange. For other regular relational operator, as Julian pointed, Calcite-831 will have the optimization rules and Drill could re-use them once Calcite-831 is done (Drill-636 was opened for those cases).